Fixes #247

STRd6 opened the pull request.

edemaine commented on source/parser.hera:
```diff
@@ -834,7 +834,9 @@ ObjectBindingPattern
-  BindingProperty*:props BindingRestProperty?:rest BindingProperty*:after ->
+  BindingPropertyList*:props BindingRestProperty?:rest BindingPropertyList*:after ->
+    props = props.flat()
+    after = after.flat()
```
It seems like `BindingPropertyList` already has a `+` so each `BindingPropertyList*` could be replaced by `BindingPropertyList?`, right? And then you could omit the `.flat()`.

STRd6 replied:
This needs the `*` to absorb multiple lines of `Nested`.

edemaine replied:
Maybe I'm missing something, but it seems like (at least in this case) there's no `Nested` being called, it's just a rule with `+` which can't actually consume `EOS`.

edemaine commented on source/parser.hera:
```diff
@@ -906,6 +917,8 @@ NestedBindingProperties
   PushIndent NestedBindingProperty*:props NestedBindingRestProperty?:rest NestedBindingProperty*:after PopIndent ->
     if (!(props.length || rest)) return $skip
+    props = props.flat()
+    after = after.flat()
```
I think each `NestedBindingProperty*` should be `NestedPropertyList?`, given that there's no longer a delimiter in `BindingProperty`? (and omit `.flat()`)

edemaine commented on source/parser.hera:
```diff
@@ -919,20 +932,21 @@ NestedBindingProperties
 NestedBindingProperty
-  Nested:indent BindingProperty:prop ->
-    return {
+  Nested:ws ( BindingProperty ObjectPropertyDelimiter )+:inlineProps ->
```
Seems like `( BindingProperty ObjectPropertyDelimiter )+` should be replaced by equivalent `BindingPropertyList`. Oh, you're doing something different in the code block. But maybe just the whitespace insertion could be done in this code block, and avoid duplicating code?

STRd6 commented:
Things were getting cray cray so I decided to just refactor it into a less crazy way.

- Moved BindingRestProperty into BindingProperty
  - No longer enforcing exactly one rest property at the parser level (letting TSServer handle the error notices)
  - Moved delimiter checking into container rules
- Moved all reordering logic into `module.reorderBindingRestProperty`

edemaine commented on source/parser.hera:
```diff
@@ -5688,20 +5684,49 @@ Init
 
-        rest = {...rest, children: [...rest.children.slice(0, -1), lastDelim]}
-        after = [...after.slice(0, -1), {...lastAfterProp, children: [...lastAfterChildren.slice(0, -1), restDelim]}]
+      if (restCount === 1) {
```
If you wanted to, you could throw a parse error when `restCount > 1`. Maybe beneficial? But not a big deal.

edemaine reviewed (approved):
Oh wow, this is so much cleaner! Looks good to me; one idea for better feedback.
